File: backend/search_engine.py
from typing import List, Dict, Tuple
import math
import re
from collections import Counter
from standards_data import IndianStandard

# BM25 parameters
K1 = 1.5
B = 0.75


def tokenize(text: str) -> List[str]:
    if not text:
        return []
    text = text.lower()
    text = re.sub(r'[^\w\s]', ' ', text)
    tokens = [t for t in text.split() if len(t) > 1]
    return tokens


def get_document_text(std: IndianStandard) -> str:
    parts = [
        std.id, std.title, std.category, std.subcategory, std.scope,
        std.description, ' '.join(std.keywords),
        ' '.join(std.product_examples),
        std.certification_scheme, std.applicability_reason,
        ' '.join(std.what_it_covers)
    ]
    return ' '.join(parts)


def build_df(documents: List[List[str]]) -> Dict[str, int]:
    df = {}
    for doc in documents:
        for term in set(doc):
            df[term] = df.get(term, 0) + 1
    return df


def bm25_score(query_terms, doc_terms, df, total_docs, avg_doc_len):
    score = 0.0
    doc_len = len(doc_terms)
    tf = Counter(doc_terms)
    for term in query_terms:
        freq = tf.get(term, 0)
        if freq == 0:
            continue
        doc_count = df.get(term, 0)
        idf = math.log((total_docs - doc_count + 0.5) / (doc_count + 0.5) + 1)
        tf_norm = (freq * (K1 + 1)) / (freq + K1 * (1 - B + B * (doc_len / avg_doc_len)))
        score += idf * tf_norm
    return score


def keyword_score(query: str, std: IndianStandard) -> float:
    q = query.lower()
    score = 0.0
    if std.category.lower() in q:
        score += 3
    if std.subcategory.lower() in q:
        score += 2
    for ex in std.product_examples:
        if ex.lower() in q:
            score += 4
    for kw in std.keywords:
        if kw.lower() in q:
            score += 2
    if std.id.lower() in q:
        score += 5
    return score


def semantic_score(query: str, std: IndianStandard) -> float:
    q_terms = set(tokenize(query))
    d_terms = set(tokenize(get_document_text(std)))
    overlap = len(q_terms & d_terms)
    return overlap / len(q_terms) if q_terms else 0.0


class SearchResult:
    def __init__(self, standard, bm25, keyword, semantic, combined, reason, confidence):
        self.standard = standard
        self.bm25_score = bm25
        self.keyword_score = keyword
        self.semantic_score = semantic
        self.combined_score = combined
        self.match_reason = reason
        self.confidence = confidence

    def to_dict(self):
        return {
            'standard': self.standard.to_dict(),
            'bm25_score': round(self.bm25_score, 3),
            'keyword_score': round(self.keyword_score, 3),
            'semantic_score': round(self.semantic_score, 3),
            'combined_score': round(self.combined_score, 3),
            'match_reason': self.match_reason,
            'confidence': self.confidence,
        }


def generate_match_reason(query: str, std: IndianStandard) -> str:
    q = query.lower()
    for ex in std.product_examples:
        if ex.lower() in q:
            return f'Your product matches "{ex}" which falls under the scope of {std.id}.'
    if std.category.lower() in q:
        return f'Your product belongs to the "{std.category}" category, which is covered by {std.id}.'
    for kw in std.keywords:
        if kw.lower() in q:
            return f'The term "{kw}" in your description relates to {std.id} — {std.title}.'
    return f'Based on your product description, this standard may be relevant to your product category.'


def search_standards(query: str, standards: List[IndianStandard], max_results: int = 10) -> List[SearchResult]:
    q_terms = tokenize(query)
    documents = [tokenize(get_document_text(s)) for s in standards]
    df = build_df(documents)
    total = len(standards)
    avg_len = sum(len(d) for d in documents) / total if total else 0

    results = []
    for std in standards:
        doc_terms = tokenize(get_document_text(std))
        bm25 = bm25_score(q_terms, doc_terms, df, total, avg_len)
        kw = keyword_score(query, std)
        sem = semantic_score(query, std)
        combined = bm25 * 0.4 + kw * 0.35 + sem * 0.25

        if combined > 0.5 or kw > 3:
            confidence = 'high' if combined > 3 else ('medium' if combined > 1.5 else 'low')
            results.append(SearchResult(std, bm25, kw, sem, combined, generate_match_reason(query, std), confidence))

    results.sort(key=lambda r: r.combined_score, reverse=True)
    return results[:max_results]


def reciprocal_rank_fusion(ranked_lists, k: int = 60):
    scores = {}
    for lst in ranked_lists:
        for rank, item in enumerate(lst):
            rrf = 1.0 / (k + rank + 1)
            std_id = item['standard']['id'] if isinstance(item, dict) else item.standard.id
            if std_id in scores:
                scores[std_id]['score'] += rrf
            else:
                scores[std_id] = {'standard': item if isinstance(item, dict) else item.to_dict(), 'score': rrf}
    return sorted(scores.values(), key=lambda x: x['score'], reverse=True)

File: src/utils/searchEngine.ts
import { IndianStandard, PRODUCT_STANDARD_MAPPINGS } from '../data/standards';

const K1 = 1.5;
const B = 0.75;

function tokenize(text: string): string[] {
  return text.toLowerCase()
    .replace(/[^\w\s]/g, ' ')
    .split(/\s+/)
    .filter(t => t.length > 1);
}

function buildDF(documents: string[][]): Map<string, number> {
  const df = new Map<string, number>();
  for (const doc of documents) {
    const uniqueTerms = new Set(doc);
    for (const term of uniqueTerms) {
      df.set(term, (df.get(term) || 0) + 1);
    }
  }
  return df;
}

function bm25Score(
  queryTerms: string[],
  documentTerms: string[],
  df: Map<string, number>,
  totalDocs: number,
  avgDocLength: number
): number {
  let score = 0;
  const docLength = documentTerms.length;
  const termFreq = new Map<string, number>();

  for (const term of documentTerms) {
    termFreq.set(term, (termFreq.get(term) || 0) + 1);
  }

  for (const term of queryTerms) {
    const tf = termFreq.get(term) || 0;
    const docCount = df.get(term) || 0;

    if (tf === 0) continue;

    const idf = Math.log((totalDocs - docCount + 0.5) / (docCount + 0.5) + 1);
    const tfNorm = (tf * (K1 + 1)) / (tf + K1 * (1 - B + B * (docLength / avgDocLength)));

    score += idf * tfNorm;
  }

  return score;
}

function getDocumentText(std: IndianStandard): string {
  return [
    std.id,
    std.title,
    std.category,
    std.subcategory,
    std.scope,
    std.description,
    std.keywords.join(' '),
    std.productExamples.join(' '),
    std.certificationScheme,
    std.applicabilityReason,
    std.whatItCovers.join(' '),
  ].join(' ');
}

export interface SearchResult {
  standard: IndianStandard;
  bm25Score: number;
  keywordScore: number;
  semanticScore: number;
  combinedScore: number;
  matchReason: string;
  confidence: 'high' | 'medium' | 'low';
}

function keywordScore(query: string, standard: IndianStandard): number {
  const queryLower = query.toLowerCase();
  let score = 0;

  if (queryLower.includes(standard.category.toLowerCase())) score += 3;
  if (queryLower.includes(standard.subcategory.toLowerCase())) score += 2;

  for (const example of standard.productExamples) {
    if (queryLower.includes(example.toLowerCase())) score += 4;
  }

  for (const keyword of standard.keywords) {
    if (queryLower.includes(keyword.toLowerCase())) score += 2;
  }

  if (queryLower.includes(standard.id.toLowerCase())) score += 5;

  return score;
}

function semanticScore(query: string, standard: IndianStandard): number {
  const queryTerms = new Set(tokenize(query));
  const docTerms = new Set(tokenize(getDocumentText(standard)));

  let overlap = 0;
  for (const term of queryTerms) {
    if (docTerms.has(term)) overlap++;
  }

  return queryTerms.size > 0 ? overlap / queryTerms.size : 0;
}

export function searchStandards(
  query: string,
  standards: IndianStandard[],
  maxResults: number = 10
): SearchResult[] {
  const queryTerms = tokenize(query);
  const documents = standards.map(s => tokenize(getDocumentText(s)));
  const df = buildDF(documents);
  const totalDocs = standards.length;
  const avgDocLength = documents.reduce((sum, d) => sum + d.length, 0) / totalDocs;

  const results: SearchResult[] = [];

  for (let i = 0; i < standards.length; i++) {
    const bm25 = bm25Score(queryTerms, documents[i], df, totalDocs, avgDocLength);
    const kw = keywordScore(query, standards[i]);
    const sem = semanticScore(query, standards[i]);

    const combined = bm25 * 0.4 + kw * 0.35 + sem * 0.25;

    if (combined > 0.5 || kw > 3) {
      const confidence = combined > 3 ? 'high' : combined > 1.5 ? 'medium' : 'low';

      results.push({
        standard: standards[i],
        bm25Score: bm25,
        keywordScore: kw,
        semanticScore: sem,
        combinedScore: combined,
        matchReason: generateMatchReason(query, standards[i]),
        confidence,
      });
    }
  }

  results.sort((a, b) => b.combinedScore - a.combinedScore);

  return results.slice(0, maxResults);
}

function generateMatchReason(query: string, standard: IndianStandard): string {
  const queryLower = query.toLowerCase();

  for (const example of standard.productExamples) {
    if (queryLower.includes(example.toLowerCase())) {
      return `Your product matches "${example}" which falls under the scope of ${standard.id}.`;
    }
  }

  if (queryLower.includes(standard.category.toLowerCase())) {
    return `Your product belongs to the "${standard.category}" category, which is covered by ${standard.id}.`;
  }

  for (const keyword of standard.keywords) {
    if (queryLower.includes(keyword.toLowerCase())) {
      return `The term "${keyword}" in your description relates to ${standard.id} - ${standard.title}.`;
    }
  }

  return `Based on your product description, this standard may be relevant to your product category.`;
}

export function lookupProductMapping(query: string) {
  const queryLower = query.toLowerCase();

  for (const mapping of PRODUCT_STANDARD_MAPPINGS) {
    const mappingWords = mapping.productQuery.toLowerCase().split(' ');
    const queryWords = queryLower.split(' ');

    const overlap = mappingWords.filter(w => queryWords.includes(w)).length;
    const threshold = Math.min(mappingWords.length, queryWords.length) * 0.4;

    if (overlap >= threshold) {
      return mapping;
    }
  }

  return null;
}

export function validateISCode(code: string, whitelist: Set<string>): boolean {
  const normalized = code.trim().toUpperCase().replace(/\s+/g, ' ');
  return whitelist.has(normalized);
}

export function extractISCodes(text: string): string[] {
  const pattern = /IS\s*\d+(\.\d+)?/gi;
  const matches = text.match(pattern) || [];
  return [...new Set(matches.map(m => m.trim().toUpperCase()))];
}

export function reciprocalRankFusion(
  rankedLists: { standard: IndianStandard; score: number }[][],
  k: number = 60
): { standard: IndianStandard; rrfScore: number }[] {
  const scores = new Map<string, { standard: IndianStandard; score: number }>();

  for (const list of rankedLists) {
    list.forEach((item, rank) => {
      const rrfScore = 1 / (k + rank + 1);
      const existing = scores.get(item.standard.id);
      if (existing) {
        existing.score += rrfScore;
      } else {
        scores.set(item.standard.id, { standard: item.standard, score: rrfScore });
      }
    });
  }

  return Array.from(scores.values())
    .sort((a, b) => b.score - a.score)
    .map(item => ({ standard: item.standard, rrfScore: item.score }));
}
